### PR TITLE
update HttpSemanticConventions for Instrumentation.Http

### PR DESCRIPTION
**Pull request by TimothyMothra**
Design discussion issue #4484

## Changes
- add new attributes to `SemanticConventions`
- update `Instrumentation.Http`
  - `HttpHandlerDiagnosticListener`
  - `HttpHandlerMetricsDiagnosticListener`

Please provide a brief description of the changes here.

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [ ] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

@vishweshbankwar please review

### CODE REVIEW

**Comment by vishweshbankwar** [comment]
FYI - we will also need to update this https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.Http/Implementation/HttpWebRequestActivitySource.netfx.cs (separate PR)

**Comment by TimothyMothra** [comment]
> FYI - we will also need to update this https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.Http/Implementation/HttpWebRequestActivitySource.netfx.cs (separate PR)

Sorry, I missed it because I was only looking at the "Listener" classes.
Any reason to not include it in this PR?

File: src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs [hunk 1]
```diff
@@ -55,19 +55,43 @@ public override void OnEventWritten(string name, object payload)
...
+                        }
                     }
 
-                    if (this.stopResponseFetcher.TryFetch(payload, out HttpResponseMessage response) && response != null)
+                    // see the spec https://github.com/open-telemetry/semantic-conventions/blob/main/specification/trace/semantic_conventions/http.md
+                    if (this.httpSemanticConvention.HasFlag(HttpSemanticConvention.New))
                     {
-                        tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeHttpStatusCode, TelemetryHelper.GetBoxedStatusCode(response.StatusCode)));
+                        tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeHttpRequestMethod, HttpTagHelper.GetNameForHttpMethod(request.Method)));
+                        tags.Add(new KeyValuePair<string, object>(SemanticConventions.AttributeHttpRequestMethod, request.RequestUri.Scheme));
```
Comment (vishweshbankwar) [inline src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs:82]: typo?

File: src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs [hunk 2]
```diff
@@ -162,16 +162,44 @@ public void OnStartActivity(Activity activity, object payload)
...
+                        activity.SetTag(SemanticConventions.AttributeNetPeerPort, request.RequestUri.Port);
+                    }
+
+                    activity.SetTag(SemanticConventions.AttributeHttpUrl, HttpTagHelper.GetUriTagValueFromRequestUri(request.RequestUri));
+                    activity.SetTag(SemanticConventions.AttributeHttpFlavor, HttpTagHelper.GetFlavorTagValueFromProtocolVersion(request.Version));
                 }
 
-                activity.SetTag(SemanticConventions.AttributeHttpUrl, HttpTagHelper.GetUriTagValueFromRequestUri(request.RequestUri));
-                activity.SetTag(SemanticConventions.AttributeHttpFlavor, HttpTagHelper.GetFlavorTagValueFromProtocolVersion(request.Version));
+                // see the spec https://github.com/open-telemetry/semantic-conventions/blob/main/specification/trace/semantic_conventions/http.md
```
Comment (utpilla) [inline src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs:180]: Update the links to point to a commit/tag.

File: src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs [hunk 3]
```diff
@@ -162,16 +162,44 @@ public void OnStartActivity(Activity activity, object payload)
...
+                    {
+                        activity.SetTag(SemanticConventions.AttributeServerPort, request.RequestUri.Port);
+                    }
+
+                    activity.SetTag(SemanticConventions.AttributeUrlFull, HttpTagHelper.GetUriTagValueFromRequestUri(request.RequestUri));
+                    activity.SetTag(SemanticConventions.AttributeNetworkProtocolVersion, HttpTagHelper.GetFlavorTagValueFromProtocolVersion(request.Version));
+
+                    if (request.Headers.TryGetValues("User-Agent", out var userAgentValues))
+                    {
+                        var userAgent = userAgentValues.FirstOrDefault();
```
Comment (utpilla) [inline src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs:196]: This would cause an allocation. Refer to how AspNetCore instrumentation handles this: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L217-L224

This should be done in a separate PR which updates the benchmarks as well.

**Review by utpilla — APPROVED** [review]
Left a comment about updating the urls to point to a commit/tag instead of main branch.

Need a CHANGELOG entry.

File: src/OpenTelemetry.Api/Internal/SemanticConventions.cs [hunk 4]
```diff
@@ -110,5 +110,15 @@ internal static class SemanticConventions
         public const string AttributeExceptionType = "exception.type";
         public const string AttributeExceptionMessage = "exception.message";
         public const string AttributeExceptionStacktrace = "exception.stacktrace";
+
+        // NEW v1.21.0 Http Semantic Conventions
```
Comment (reyang) [inline src/OpenTelemetry.Api/Internal/SemanticConventions.cs:114]: What does this mean?

Comment (TimothyMothra) [reply src/OpenTelemetry.Api/Internal/SemanticConventions.cs:114]: I was asked to separate the new attributes from the others and give it a descriptive comment.
The word "new" isn't needed here, sorry for the confusion.
I've updated this comment in the AspNetCore PR but forgot to update it here.

Comment (reyang) [reply src/OpenTelemetry.Api/Internal/SemanticConventions.cs:114]: > I was asked to separate the new attributes from the others and give it a descriptive comment.

What does v1.21.0 mean?

> The word "new" isn't needed here

👍

File: src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs (hunk 2 above)
Comment (TimothyMothra) [reply src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs:180]: done.

File: src/OpenTelemetry.Api/Internal/SemanticConventions.cs [hunk 5]
```diff
@@ -110,5 +110,15 @@ internal static class SemanticConventions
         public const string AttributeExceptionType = "exception.type";
         public const string AttributeExceptionMessage = "exception.message";
         public const string AttributeExceptionStacktrace = "exception.stacktrace";
+
+        // Http v1.21.0 https://github.com/open-telemetry/opentelemetry-specification/blob/v1.21.0/specification/trace/semantic_conventions/http.md
```
Comment (reyang) [inline src/OpenTelemetry.Api/Internal/SemanticConventions.cs:114]: FYI - v1.21.0 doesn't contain semantic convention changes https://github.com/open-telemetry/opentelemetry-specification/blob/v1.21.0/CHANGELOG.md#v1210-2023-05-09

image

Comment (TimothyMothra) [reply src/OpenTelemetry.Api/Internal/SemanticConventions.cs:114]: The **semantic-conventions** repo doesn't have any [tags ](https://github.com/open-telemetry/semantic-conventions/tags) yet so I can't link to a versioned copy of the Http document.


I discussed this with Utkarsh last week and he suggested using the v1.21.0 tag on the **opentelemetry-specification** repo.
The **opentelemetry-specification** repo does have tags and that version of the document does include all of the new attributes that we're adding in this PR.

**Review by utpilla — APPROVED** [review]
